### ISSUE TITLE
promtool: Calculate coverage data for tests

### ISSUE THREAD

**Issue by dgl** [prometheus/prometheus#8131]
# Proposal

Make it possible to calculate whether the tests cover each rule and alert.

In order to do this some basic assumptions would have to be made, for example take a simple set of rules:

- `job:up:ratio` calculates the "up ratio of a job"
- Alert `TooManyDown` alerts if `job:up:ratio < 0.5`

The input would be just the `up` metric, potentially per job. This would mean `job:up:ratio` would have a selector like `up`, this would therefore be the expected input and something like this should exist in `input_series` for the test.

Even in this simple example there's some tricky bits, is it enough if an alert test covers job:up:ratio, or would it be better to count  the rule coverage differently from the alert coverage. (Seems like reporting them separately and the user can choose to interpret them as they like would work though.)

More tricky things are `label_replace()` and regexps on `__name__`, as it would be very tricky to follow the rule dependencies there but I think this isn't a huge problem -- in most cases it should be possible to follow a rule back to the input series via selectors. Where this fails the result is coverage is reported as less than it might actually be, potentially the coverage output could warn with something like "regexp selector seen on __name__, coverage reported may be less than actual".

Coverage would need to be calculated per rule file, but not per test file (i.e. don't enforce anything like tests for a particular rule file have to be in a single test file -- we consider all the arguments to `promtool test rules` as a "suite"). However inputs would be per test file (i.e. if one test treats the output of recording rules from another file not under test, that would be fine).

I'm imaging the output could be something like:

```
rule-file.yaml:
    50% rule coverage
    50% alert coverage
```

**Use case. Why is this important?**

It would be good to be able to get an idea of how well unit tests cover the rules they are supposed to be testing. This brings PromQL closer to something you'd expect of a fully featured language.

I'm willing to work on this, maybe will prototype and see how it goes. I think it makes sense in promtool as it needs knowledge of how PromQL executes and can't quite be pushed to a PromQL language server (although maybe there are integration opportunities there, later).

**Comment by brian-brazil** [comment]
> It would be good to be able to get an idea of how well unit tests cover the rules they are supposed to be testing.

I'm a bit wary of this, as it would encourage getting 100% test coverage rather than spending time testing what is actually useful (which is usually only the alerts). For example if there are a pile of copy&pasted rules doing the same basic aggregation to different non-critical metrics, are writing and maintaining exhaustive unittests the best use of someone's time?

I'd prefer to encourage testing complex and non-trivial rules, over writing low-value unittests for simple rules.

**Comment by PicksUpTheGauntlet** [comment]
I agree with this suggestion. Even with simple rules, I think unit tests can be valuable. I'm open to be corrected but as it stands, please consider this as an example: 

1. Many targets are queried under a generic job and you've created a single alert used for a specific event. 
2. As your incident management becomes more mature, your team starts creating runbooks and you realize that one of the targets in the job needs its own alert description, so a link to the runbook can be included in the notification. 
3. You're required to now make a separate alert which specifies the target in the expression, from within the same job as above. You believe the alert to be simple to put in place (as it should be) and you do so without a unit test. 
5. The next time the target is experiencing an issue, 2 alerts are triggered because you've forgot to add a `!=target` condition in the first alert's expression, to ignore the target you configured in step 3.

By configuring a `input_series` (at the very least) and an `alert_rule_test`  specifically for this job and target, we could have been advised through `promtool test rules` that a specific series was triggering more than 1 alert in our "alert_rules.yml" file.

While there could be another layer of abstraction handling this documentation scenario, I believe this justification holds a bit of weight. Maybe it's not documentation but rather a change in an alert label that routes the alert to a different team. _In other words, for any reason that requires you to cherry-pick targets out of generic alerts. In an ideal world, exceptions wouldn't exist..._ 

I hope I've demonstrated the scenario well enough, to help this be considered and I'm open to the idea that maybe I've presented an anti-pattern...

**Comment by stacybird** [comment]
@dgl did the prototype go anywhere on this, or didn't have a chance to take a look?  I don't see a branch or PR mentioned here.

**Comment by jnovick** [comment]
Since `promtool` does not have this feature, this is a short script I used to generate coverage. I adjusted it to be more generic and not specific to the folders I used. I did not test yhis adjusted version but it should work the same as what I did. Feel free to copy and use this for yourself until the feature is added into the tool.

``` bash
TOTAL_ALERTS=$(ls rules/*.yaml | xargs -I {} yq e '.groups[] | .rules[] | .alert' {} | sort | uniq | wc -l | sed 's/ //g' || true)
ALERT_FILE_COUNT=$(ls rules/*.yaml | wc -l | sed 's/ //g' || true)
TEST_COUNT=$( (find -L tests -type f -name "*.yaml" -exec yq e '.tests | length' {} \; | awk '{s+=$1} END {print s}' 2> /dev/null || echo 0 ) | tr -d "\n" )
TEST_FILE_COUNT=$(find -L tests -type f -name "*.yaml" | wc -l | sed 's/ //g' || true)
TESTED_ALERTS=$(find -L tests -type f -name "*.yaml" -exec yq e '.tests[] | .alert_rule_test[] | .alertname' {} \; | sort | uniq | wc -l | sed 's/ //g' || true)
TEST_COVERAGE=$( [[ "$TOTAL_ALERTS" -eq 0 ]] && echo NaN || ( echo $(( $TESTED_ALERTS * 10000 / $TOTAL_ALERTS )) | sed 's/\(.*\)\([[:digit:]][[:digit:]]\)/\1.\2/' ) )

echo "$TOTAL_ALERTS alerts were found in $ALERT_FILE_COUNT files."
echo "$TEST_COUNT tests were found in $TEST_FILE_COUNT files."
echo "$TESTED_ALERTS/$TOTAL_ALERTS ($TEST_COVERAGE%) alerts have at least one test. (This does not mean the tests are good quality)" 
```

**Comment by beorn7** [comment]
Hello from the bug scrub.

@dgl this issue is quite old by now. Do you think it's still worth pursuing?
